tools/cmaketobzl/cmaketobzl.go

```diff
@@ -0,0 +1,389 @@
+func (e *eval) excludePath(dirpath string) bool {
+	return e.o.excludePath != nil && e.o.excludePath(dirpath)
+}
+
+// dispatch evaluates the next command from cmds and returns a new dispatchFunc for handling the remainder.
+func (e *eval) dispatch(cmds *commandList) (dispatchFunc, error) {
+	name := string(cmds.Head().Name)
+	if e.shouldPrint(name) {
+		e.PrintCommand(cmds.Head())
+	}
+
+	switch name {
+	// TODO(shahms): Actually process these.
+	case "if", "function", "foreach", "macro":
+		counter := newCounter(name)
+		for counter.Count(name) && cmds.Advance() {
+			name = string(cmds.Head().Name)
+		}
+		return e.dispatch, nil
+	// TODO(shahms): Support setting values.
+	case "set":
+	}
+
+	if e.shouldAdd(name) {
+		args := cmds.Head().Arguments.Eval(e.v)
+		if len(args) != 1 {
+			return nil, fmt.Errorf("invalid number of arguments to directory command %s", cmds.Head().Pos)
+		}
+		if !e.excludePath(args[0]) {
+			if err := e.AddSubdirectory(cmds.Head().Arguments.Eval(e.v)[0]); err != nil {
+				return nil, err
+			}
+		}
+	}
+	cmds.Advance()
+	return e.dispatch, nil
+}
+
+// AddSubdirectory recurses into the directory specified by dirpath and evaluates the CMakeLists.txt contained therein.
+func (e *eval) AddSubdirectory(dirpath string) error {
+	if err := e.enterDirectory(dirpath); err != nil {
+		return err
+	}
+	file, err := e.parseFile(path.Join(path.Join(e.path...), "CMakeLists.txt"))
+	if err != nil {
+		return err
+	}
+
+	cmds := commandList(file.Commands)
+	dispatch := e.dispatch
+	for {
+		dispatch, err := dispatch(&cmds)
+		if err != nil {
+			return err
+		}
+		if len(cmds) == 0 || dispatch == nil {
+			break
+		}
+
+	}
+	return e.exitDirectory(dirpath)
+}
+
+// CurrentDirectory returns the project-rooted path currently being traversed.
+func (e *eval) CurrentDirectory() string {
+	return "/" + path.Join(e.path[1:]...)
+}
+
+// enterDirectory pushes a new directory onto the stack, setting up necessary state, etc.
+func (e *eval) enterDirectory(dirpath string) error {
+	if err := e.w.PushDirectory(dirpath); err != nil {
+		return err
+	}
+	e.v.Push()
+	e.path = append(e.path, dirpath)
+	e.v.Set("CMAKE_CURRENT_SOURCE_DIR", e.CurrentDirectory())
+	e.v.Set("CMAKE_CURRENT_BINARY_DIR", e.CurrentDirectory())
+	return nil
+}
+
+// exitDirectory pops the most recently entered directory off the stack.
+func (e *eval) exitDirectory(path string) error {
+	e.v.Pop()
+	e.path = e.path[:len(e.path)-1]
+	tail, err := e.w.PopDirectory()
+	if tail != path {
+		return fmt.Errorf("unexpected directory state %v != %v", tail, path)
+	}
+	return err
+}
+
+// PrintCommand writes the given command to the configured StarlarkWriter.
+func (e *eval) PrintCommand(command *ast.CommandInvocation) error {
+	return e.w.WriteCommand(string(command.Name), command.Arguments.Eval(e.v)...)
+}
+
+// Path is a slice of string segments, representing a filesystem path.
+type Path []string
+
+// Split cleans and splits the /-delimited filesystem path.
```

Comment:
I'm not sure whether we care about platform compatibility here, but if so we should probably include a call to https://godoc.org/path/filepath#ToSlash if we want to assume the delimiter is slash.

Reply:
Fair enough, done.